shell/platform/darwin/macos/framework/Source/FlutterExternalTextureGL.mm

```diff
@@ -0,0 +1,87 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterExternalTextureGL.h"
+
+#import <AppKit/AppKit.h>
+#import <CoreVideo/CoreVideo.h>
+#import <OpenGL/gl.h>
+
+static void OnCVOpenGLTextureRelease(CVOpenGLTextureRef cvOpenGLTexture) {
+  CVOpenGLTextureRelease(cvOpenGLTexture);
+}
+
+@implementation FlutterExternalTextureGL {
+  /**
+   * OpenGL texture cache.
+   */
+  CVOpenGLTextureCacheRef _openGLTextureCache;
+  /**
+   * User side texture object, used to copy pixel buffer.
+   */
+  id<FlutterTexture> _texture;
+}
+
+- (instancetype)initWithFlutterTexture:(id<FlutterTexture>)texture {
+  self = [super init];
+  if (self) {
+    _texture = texture;
+  }
+  return self;
+}
+
+- (int64_t)textureID {
+  return reinterpret_cast<int64_t>(self);
+}
+
+- (BOOL)populateTextureWithWidth:(size_t)width
```

Comment:
You are not using the `width` and the `height` argument in this method at all!

The way this API works it that the engine tells the embedder what size it needs the texture to be. If the texture is not of that exact size, the engine will attempt to clamp or resize the texture to get something usable. The clamping behavior is what you seem to be running into.

There is information being thrown away in this method. That is, when you ask the texture to `copyPixelBuffer`, you are not giving it any indication about what the size of the pixel buffer should be. This method acts as if size does not matter. When, in fact, the engine care about the size and it is asking you for a texture of that size. This method is just not using it.

So now you have a texture of some unspecified size that you need to give to the engine because you didn't use the dimensions the engine asked of the embedder. To solve this issue, we now need to resort to the expensive device to host transfer so we can resize on the CPU before transferring that information back to the device.

If we don't throw away the information the engine gives us, we can be a whole lot faster. This can be done in one of the following ways:

* Augment `copyPixelBuffer` to take the `width` and the `height`. This way, plugins will know what texture the engine expects and can create texture of the right size to begin with.
* If it is unreasonable for plugins to expect to create textures of the correct size, we can still make the `FlutterOpenGLTexture` take a `width` and `height` parameter (subject to ABI restrictions I mentioned below). That way, we give the plugins the best chance to match texture sizes. If however, the size the engine expects is not the same as the size the plugin gives back to us, instead of implementing the resize via a software render pass, we can achieve the same by applying a transformation to the draw on the GPU. This can be modified by pushing a [CTM on the canvas here](https://github.com/flutter/engine/blob/49815128157d80803da9f36c69493b2a09aacd03/shell/platform/embedder/embedder_external_texture_gl.cc#L34).

Whatever you choose, let do as many operations on the GPU as possible by utilizing all information the engine give to the embedder.